ddtrace/_trace/_span_pointers.py

```diff
@@ -0,0 +1,92 @@
+# A private API for now.
+
+from enum import Enum
+from hashlib import sha256
+from typing import Any
+from typing import Dict
+from typing import NamedTuple
+from typing import Optional
+
+from ddtrace._trace._span_link import SpanLink
+from ddtrace.internal.logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+class _SpanPointerDirection(Enum):
+    UPSTREAM = "u"
+    DOWNSTREAM = "d"
+
+
+class _SpanPointerDescription(NamedTuple):
+    # Not to be confused with _SpanPointer. This class describes the parameters
+    # required to attach a span pointer to a Span. It lets us decouple code
+    # that calculates span pointers from code that actually attaches them to
+    # the right Span.
+
+    pointer_kind: str
+    pointer_direction: _SpanPointerDirection
+    pointer_hash: str
+    extra_attributes: Dict[str, Any]
+
+
+_SPAN_LINK_KIND_SPAN_POINTER = "span-pointer"
+_SPAN_POINTER_SPAN_LINK_TRACE_ID = 0
+_SPAN_POINTER_SPAN_LINK_SPAN_ID = 0
+
+
+class _SpanPointer(SpanLink):
+    # The class to privately store span pointers and serialize them alongside
+    # Span Links. The span pointer methods on Span objects should be used
+    # instead of instantiating this class directly.
+
+    def __init__(
+        self,
+        pointer_kind: str,
+        pointer_direction: _SpanPointerDirection,
+        pointer_hash: str,
+        extra_attributes: Optional[Dict[str, Any]] = None,
+    ):
+        super().__init__(
+            trace_id=_SPAN_POINTER_SPAN_LINK_TRACE_ID,
+            span_id=_SPAN_POINTER_SPAN_LINK_SPAN_ID,
+            attributes={
+                "ptr.kind": pointer_kind,
+                "ptr.dir": pointer_direction.value,
+                "ptr.hash": pointer_hash,
+                **(extra_attributes or {}),
+            },
+        )
+
+        self.kind = _SPAN_LINK_KIND_SPAN_POINTER
+
+    def __post_init__(self):
+        if self.trace_id != _SPAN_POINTER_SPAN_LINK_TRACE_ID:
+            log.warning(
+                "span pointer trace_id must be %d, found %d",
+                _SPAN_POINTER_SPAN_LINK_TRACE_ID,
+                self.trace_id,
+            )
+            self.trace_id = _SPAN_POINTER_SPAN_LINK_TRACE_ID
+
+        if self.span_id != _SPAN_POINTER_SPAN_LINK_SPAN_ID:
+            log.warning(
+                "span pointer span_id must be %d, found %d",
+                _SPAN_POINTER_SPAN_LINK_SPAN_ID,
+                self.span_id,
+            )
+            self.span_id = _SPAN_POINTER_SPAN_LINK_SPAN_ID
+
+
+def _standard_hashing_function(*elements: bytes) -> str:
+    separator = b"|"
+    bits_per_hex_digit = 4
+    desired_bits = 128
+    hex_digits = desired_bits // bits_per_hex_digit
+
+    hex_digest = sha256(separator.join(elements)).hexdigest()
+    if len(hex_digest) < hex_digits:
+        raise Exception(f"the hash is smaller than required: {len(hex_digest)} < {hex_digits}")
```

Comment:
hmm. anyone have an opinion on what this exception should be? this is *technically* impossible since the sha256 hash is guaranteed to have more than the requested number of digits in its hex digest. but this is a bit of self-documenting code that should help us avoid potential problems in the future. on the other other hand once this standard hashing function is written, it is almost certainly not going to be changing. so maybe we just leave this and the other calculations in here out of the function and just write some tests?

Reply:
no more exceptoin, anyway.